// librato-node/lib/aggregator.js
// Generated by CoffeeScript 1.9.1
(function() {
  var Aggregator, max, min, sum;

  sum = function(values) {
    return values.reduce(function(a, b) {
      return a + b;
    });
  };

  max = function(values) {
    return values.reduce(function(a, b) {
      return Math.max(a, b);
    });
  };

  min = function(values) {
    return values.reduce(function(a, b) {
      return Math.min(a, b);
    });
  };

  Aggregator = (function() {
    function Aggregator() {
      this.cache = {};
    }

    Aggregator.prototype.flushTo = function(queue) {
      var name, ref, results, values;
      ref = this.cache;
      results = [];
      for (name in ref) {
        values = ref[name];
        values.sort();
        queue.push({
          name: name,
          count: values.length,
          sum: sum(values),
          max: max(values),
          min: min(values),
          sum_squares: sum(values.map(function(value) {
            return Math.pow(value, 2);
          }))
        });
        results.push(delete this.cache[name]);
      }
      return results;
    };

    Aggregator.prototype.timing = function(name, value) {
      var base;
      return ((base = this.cache)[name] != null ? base[name] : base[name] = []).push(value);
    };

    return Aggregator;

  })();

  module.exports = Aggregator;

}).call(this);
